src/allocator.f90

```diff
@@ -69,16 +83,47 @@ module m_allocator
 
 contains
 
-   function field_constructor(dims, next, id) result(m)
-      integer, intent(in) :: dims(3), id
+   function field_init(ngrid, next, id) result(f)
+      integer, intent(in) :: ngrid, id
       type(field_t), pointer, intent(in) :: next
-      type(field_t) :: m
+      type(field_t) :: f
+
+      allocate (f%p_data(ngrid))
+      f%refcount = 0
+      f%next => next
+      f%id = id
+   end function field_init
+
+   subroutine set_shape(self, dims)
+      implicit none
+
+      class(field_t) :: self
+      integer, intent(in) :: dims(3)
+
+      self%data(1:dims(1), 1:dims(2), 1:dims(3)) => self%p_data
+
+   end subroutine set_shape
 
-      allocate (m%data(dims(1), dims(2), dims(3)))
-      m%refcount = 0
-      m%next => next
-      m%id = id
-   end function field_constructor
+   function allocator_init(nx, ny, nz, sz) result(allocator)
+      integer, intent(in) :: nx, ny, nz, sz
+      type(allocator_t) :: allocator
+
+      integer :: nx_padded, ny_padded, nz_padded
+
+      ! Apply padding based on sz
+      nx_padded = nx - 1 + mod(-(nx - 1), sz) + sz
+      ny_padded = ny - 1 + mod(-(ny - 1), sz) + sz
+      ! Current reorder functions do not require a padding in z-direction.
+      nz_padded = nz
+
+      allocator%ngrid = nx_padded*ny_padded*nz_padded
+      allocator%sz = sz
+
+      allocator%xdims = [sz, nx_padded, ny_padded*nz_padded/sz]
+      allocator%ydims = [sz, ny_padded, nx_padded*nz_padded/sz]
+      allocator%zdims = [sz, nz_padded, nx_padded*ny_padded/sz]
+      allocator%cdims = [nx_padded, ny_padded, nz_padded]
```

Comment:
I would mention in the names `xdims`/`ydims`/`zdims` that these are the padded dimensions. Also related to #77, at some point we should maybe think about where these are available, and how they are accessed.